src/funcx_common/task_storage/chain.py

```diff
@@ -0,0 +1,71 @@
+import typing as t
+
+from ..tasks import TaskProtocol
+from .base import StorageException, TaskStorage
+
+
+class ChainedTaskStorage(TaskStorage):
+    """
+    Chained storage combines multiple other concrete storage options.
+
+    The storage methods are used on a "first-come, first-served" basis, with the
+    intended usage pattern being
+
+    >>> slow_storage = ...
+    >>> fast_storage = ...
+    >>> task_storage = ChainedTaskStorage(fast_storage, slow_storage)
+
+    In this scenario, the fast storage will be tried first for both the store and get
+    operations.
+
+    If the fast storage rejects a storage operation, the slow storage will be tried. A
+    failure (exception) would be raised immediately, however.
+
+    The assumption here is that the fast storage responds quickly enough that it can be
+    checked first in all cases, even without additional context for the call.
+    """
+
+    storage_id = "ChainedTaskStorage"
+
+    def __init__(self, *storages: TaskStorage) -> None:
+        self.storages = list(storages)
+        self.backward_compatible_storage = None
+        for storage in storages:
+            if storage.backward_compatible is True:
+                self.backward_compatible_storage = storage
+                break
+
+    def store_result(self, task: TaskProtocol, result: str) -> bool:
+        exception_stack = []
+        for store in self.storages:
+            try:
+                return store.store_result(task, result)
+            except StorageException as e:
+                exception_stack.append(e)
+
+        raise StorageException(
+            f"All storage methods failed to store data: {exception_stack}"
+        )
+
+    def get_result(self, task: TaskProtocol) -> t.Optional[str]:
+        # Return None if result and reference is missing.
+        if not task.result and not task.result_reference:
+            return None
+        # Add backward compatibility with RedisStorage
+        # v0.3.3 and prior would only set task.result
+        if task.result:
+            if self.backward_compatible_storage is not None:
+                return self.backward_compatible_storage.get_result(task)
+
+        # In v0.3.4+ a result_reference is always set when the result is
+        # stored. The reference indicated the storage mechanism used.
+        if task.result_reference:
+            for storage in self.storages:
+                t_storage_id = task.result_reference["storage_id"]
+                if storage.storage_id == t_storage_id:
+                    try:
+                        return storage.get_result(task)
+                    except StorageException:
+                        pass
```

Comment:
This strikes me as very odd logic. It's basically saying that we could make `ChainedTaskStorage(store1, store2, ...)` or something else, and then we'll just try them in series, discarding any missing data, for any that are of the appropriate type?

I think it would be better to just stop after the first match.

Reply:
Let's address this later.  A lot of the logic here will *very* likely change in the future.  In reality, I would be happy with all of this being implemented as a single class that just checks if `task.result is not None` and returns that, otherwise fetches it from the s3 url in `task.result_reference`.  The chaining logic in general is some premature optimization.  The fact that `task.result_reference` is a dictionary with a key `storage_id` means traversing some code defined hierarchy is a bit useless.  There can only be on valid storage referenced in that data structure.  We either need the ability to reference multiple storage ids or we should just use a direct registry or factory patter to look up / instantiate the storage we need.  This is functional for the moment though, so let's unblock other work and revisit this later.

Reply:
This is here to *only* work with the weirdness of tests where we have chains of the the same storage (chain(redis, redis)), which I think makes no sense.